syntax = "proto3";

package robot;
import "google/protobuf/empty.proto";

service RobotService {
  rpc SendState (GameState) returns (Actions);
  rpc Act (GameState) returns (Actions);

  rpc OnEvent (Event) returns (google.protobuf.Empty);
  rpc EndRound (RoundResult) returns (google.protobuf.Empty);
  rpc StartRound (google.protobuf.Empty) returns (google.protobuf.Empty);
}


message GameState {
  RobotState robotState = 1;
  ScannedRobotEvent enemy = 2;
}
message RoundResult{
  enum Reason{
    UNKNOWN = 0;
    WIN = 1;
    LOSS = 2;
  }
  Reason reason = 1;
}



message Event {
  oneof eventType {
    BulletHitEvent bulletHit = 1;
    BulletHitBulletEvent bulletHitBullet = 2;
    BulletMissedEvent bulletMissed = 3;
    HitByBulletEvent hitByBullet = 4;
    HitRobotEvent hitRobot = 5;
    HitWallEvent hitWall = 6;
    RobotDeathEvent robotDeath = 7;
  }
}

message BulletHitEvent {
  string name = 1;
  double energy = 2;
  Bullet bullet = 3;
}

message BulletHitBulletEvent {
  Bullet bullet = 1;
  Bullet hitBullet = 2;
}

message BulletMissedEvent {
  Bullet bullet = 1;
}

message HitByBulletEvent {
  double bearing = 1;
  Bullet bullet = 2;
}

message HitRobotEvent {
  string robotName = 1;
  double bearing = 2;
  double energy = 3;
  bool atFault = 4;
}

message HitWallEvent {
  double bearing = 1;
}






message RobotDeathEvent {
  string robotName = 1;
}


message Bullet {
  double headingRadians = 1;
  double x = 2;
  double y = 3;
  double power = 4;
  string ownerName = 5;
  string victimName = 6;
  bool isActive = 7;
  int32 bulletId = 8;
}

message RobotState {
  // Robot Position & Movement
  double x = 1;
  double y = 2;
  double velocity = 3;
  double heading = 4;           // Radians

  // Gun & Radar
  double gunHeading = 5;        // Radians
  double radarHeading = 6;      // Radians
  double gunHeat = 7;
  double gunTurnRemaining = 8;  // Radians
  double radarTurnRemaining = 9; // Radians

  // Game State
  double energy = 10;
  double battleFieldWidth = 11;
  double battleFieldHeight = 12;
  int32 roundNum = 13;
  int64 time = 14;
}

message ScannedRobotEvent {
  // Enemy Position & Movement
  double x = 1;
  double y = 2;
  double velocity = 3;
  double heading = 4;       // Radians
  double bearing = 5;       // Radians
  double distance = 6;

  // Enemy Stats
  double energy = 7;

  int64 time = 8; // turn number
}



message Action {
  enum ActionType {
    MOVE_FORWARD = 0;
    MOVE_BACKWARD = 1;
    TURN_LEFT = 2;
    TURN_RIGHT = 3;
    TURN_GUN_LEFT = 4;
    TURN_GUN_RIGHT = 5;
    FIRE = 6;
    ROTATE_RADAR_LEFT = 7;
    ROTATE_RADAR_RIGHT = 8;
    DO_NOTHING = 9;
  }
  ActionType action_type = 1;
  double value = 2;
}

message Actions {
  repeated Action actions = 1;
}

